Cubical/Structures/QuotientRing.agda

```diff
@@ -0,0 +1,248 @@
+            x + (₀ + (- x))           ≡⟨ cong (λ u → x + u) (ring+-lid _) ⟩
+            x + (- x)                 ≡⟨ ring+-rinv _ ⟩
+            ₀ ∎)
+
+    translatedDifference :
+      ∀ (x a b : R)
+      → a - b ≡ (x + a) - (x + b)
+    translatedDifference x a b =
+                a - b                       ≡⟨ cong (λ u → a + u)
+                                                    (sym (ring+-lid _)) ⟩
+                (a + (₀ + (- b)))          ≡⟨ cong (λ u → a + (u + (- b)))
+                                                    (sym (ring+-rinv _)) ⟩
+                (a + ((x + (- x)) + (- b))) ≡⟨ cong (λ u → a + u)
+                                                    (sym (ring+-assoc _ _ _)) ⟩
+                (a + (x + ((- x) + (- b)))) ≡⟨ (ring+-assoc _ _ _) ⟩
+                ((a + x) + ((- x) + (- b))) ≡⟨ cong (λ u → u + ((- x) + (- b)))
+                                                    (ring+-comm _ _) ⟩
+                ((x + a) + ((- x) + (- b))) ≡⟨ cong (λ u → (x + a) + u)
+                                                    (-isDistributive _ _) ⟩
+                ((x + a) - (x + b)) ∎
+
+
+    homogenity : ∀ (x a b : R)
+                 → (a - b ∈ I)
+                 → (x + a) - (x + b) ∈ I
+    homogenity x a b p = subst (λ u → u ∈ I) (translatedDifference x a b) p
+
+
+    translate : R → R/I → R/I
+    translate x = elim
+                    (λ r → squash/)
+                    (λ y → [ x + y ])
+                    λ y y' diffrenceInIdeal
+                      → eq/ (x + y) (x + y') (homogenity x y y' diffrenceInIdeal)
+
+    isSetR/I : isSet R/I
+    isSetR/I = squash/
+    [_]/I : (a : R) → R/I
+    [ a ]/I = [ a ]
+
+    lemma : (x y a : R)
+            → x - y ∈ I
+            → [ x + a ]/I ≡ [ y + a ]/I
+    lemma x y a x-y∈I = eq/ (x + a) (y + a) (subst (λ u → u ∈ I) calculate x-y∈I)
+      where calculate : x - y ≡ (x + a) - (y + a)
+            calculate =
+                      x - y                 ≡⟨ translatedDifference a x y ⟩
+                      ((a + x) - (a + y))   ≡⟨ cong (λ u → u - (a + y)) (ring+-comm _ _) ⟩
+                      ((x + a) - (a + y))   ≡⟨ cong (λ u → (x + a) - u) (ring+-comm _ _) ⟩
+                      ((x + a) - (y + a))   ∎
+
+
+    homogenity' : (x y : R) → (x - y ∈ I)
+                  → translate x ≡ translate y
+    homogenity' x y x-y∈I i r = pointwise-equal r i
+      where
+        pointwise-equal : ∀ (u : R/I)
+                          → translate x u ≡ translate y u
+        pointwise-equal = elimProp (λ u → isSetR/I (translate x u) (translate y u))
+                                   (λ a → lemma x y a x-y∈I)
+
+    _+/I_ : R/I → R/I → R/I
+    x +/I y = (elim R/I→R/I-isSet translate homogenity' x) y
+      where
+        R/I→R/I-isSet = (λ r →  isOfHLevelΠ 2 (λ _ → squash/))
+
+    +/I-comm : (x y : R/I) → x +/I y ≡ y +/I x
+    +/I-comm = elimProp (λ x → isOfHLevelΠ 1 λ y → squash/ (x +/I y) (y +/I x))
```

Comment:
Hmm, maybe we should have a lemma for this kind of nested elimProp's?

You can use isSet\Pi as well

Reply:
I saw you discovering the lemma in another PR... but I already implemented an elimProp2 and 3. And I use it now. If the other one is already merged, I can merge the two later, when I rebase.